### ISSUE TITLE
Matrix Propagation Issues with Multi-Site and Existing Blocks

### ISSUE THREAD

**Issue by elfacht** [craftcms/cms#4731]
### Description

I have a multi-site with multiple groups. A matrix field with content blocks had the propagation setting **Save blocks to all sites the owner element is saved in** in which the entire content was created until launch. After switching to **Save blocks to other sites in the same site group** or **Only save blocks to the site they were created in** (I tried both), the blocks of all sites except the on where it was created in, had lost their blocks. With **Only save blocks to the site they were created in** the blocks appeared after switching to another site, but after saving the entry all other sites lost their blocks again. 

This btw also happens after changing **Manage relations on a per-site basis** of entries inside matrix blocks: After changing existing sites with content, all entries on others sites are gone.

On a few entries it seems to work fine, but most of them are just emptied.

### Steps to reproduce

1. Create a multi-site with groups like this: 
<img src="https://user-images.githubusercontent.com/529443/62717355-064f3300-ba04-11e9-80ae-e310cfdaa9e9.jpg" width="150"/>

2. Create a matrix field with blocks and set the propagation method to **Save blocks to all sites the owner element is saved in**
3. Create an entry with that field, save it.
4. Switch propagation method to **Save blocks to other sites in the same site group** or  **Only save blocks to the site they were created in**
5. Edit the entry again, switch to a different site than the one who created the content. The blocks are either empty on load or after saving the entry all other sites are emptied.

### Before changing the settings
![matrix-before](https://user-images.githubusercontent.com/529443/62759940-0b06fc00-ba83-11e9-865d-d736e4a09413.gif)

### After changing the settings to 'Only save blocks to the site they were created in'
![matrix-after-only](https://user-images.githubusercontent.com/529443/62759961-19edae80-ba83-11e9-9a70-e617c9859e55.gif)

This is an excerpt of `resave/matrix-blocks` I ran after the task in the CP failed:
```
    - Resaving 266 (266) ... done
    - Resaving 17679 (17679) ... error: Skipped resaving 17679 (17679) because it's a revision.
    - Resaving 10099 (10099) ... error: Skipped resaving 10099 (10099) because it's a revision.
    - Resaving 16806 (16806) ... error: Skipped resaving 16806 (16806) because it's a revision.
    - Resaving 9792 (9792) ... error: Skipped resaving 9792 (9792) because it's a revision.
    - Resaving 10804 (10804) ... error: Skipped resaving 10804 (10804) because it's a revision.
    - Resaving 6451 (6451) ... error: Skipped resaving 6451 (6451) because it's a revision.
    - Resaving 7933 (7933) ... error: Skipped resaving 7933 (7933) because it's a revision.
    - Resaving 12278 (12278) ... error: Skipped resaving 12278 (12278) because it's a revision.
    - Resaving 1656 (1656) ... done
    - Resaving 7960 (7960) ... error: Skipped resaving 7960 (7960) because it's a revision.
    - Resaving 11957 (11957) ... error: Skipped resaving 11957 (11957) because it's a revision.
    - Resaving 15523 (15523) ... error: Skipped resaving 15523 (15523) because it's a revision.
    - Resaving 21340 (21340) ... error: Skipped resaving 21340 (21340) because it's a revision.
    - Resaving 5987 (5987) ... error: Skipped resaving 5987 (5987) because it's a revision.
    - Resaving 20031 (20031) ... error: Attempting to save an element in an unsupported site.
    - Resaving 3930 (3930) ... error: Attempting to save an element in an unsupported site.
```

### Additional info

- Craft version: 3.2.9

**Comment by elfacht** [comment]
This is an excerpt of `resave/matrix-blocks` after the task in the CP failed:
```
    - Resaving 266 (266) ... done
    - Resaving 17679 (17679) ... error: Skipped resaving 17679 (17679) because it's a revision.
    - Resaving 10099 (10099) ... error: Skipped resaving 10099 (10099) because it's a revision.
    - Resaving 16806 (16806) ... error: Skipped resaving 16806 (16806) because it's a revision.
    - Resaving 9792 (9792) ... error: Skipped resaving 9792 (9792) because it's a revision.
    - Resaving 10804 (10804) ... error: Skipped resaving 10804 (10804) because it's a revision.
    - Resaving 6451 (6451) ... error: Skipped resaving 6451 (6451) because it's a revision.
    - Resaving 7933 (7933) ... error: Skipped resaving 7933 (7933) because it's a revision.
    - Resaving 12278 (12278) ... error: Skipped resaving 12278 (12278) because it's a revision.
    - Resaving 1656 (1656) ... done
    - Resaving 7960 (7960) ... error: Skipped resaving 7960 (7960) because it's a revision.
    - Resaving 11957 (11957) ... error: Skipped resaving 11957 (11957) because it's a revision.
    - Resaving 15523 (15523) ... error: Skipped resaving 15523 (15523) because it's a revision.
    - Resaving 21340 (21340) ... error: Skipped resaving 21340 (21340) because it's a revision.
    - Resaving 5987 (5987) ... error: Skipped resaving 5987 (5987) because it's a revision.
    - Resaving 20031 (20031) ... error: Attempting to save an element in an unsupported site.
    - Resaving 3930 (3930) ... error: Attempting to save an element in an unsupported site.
```

**Comment by brandonkelly** [comment]
I just followed these steps locally and things worked as expected: After step 4 a “Resaving matrix blocks” task kicked off, and when that was finished, the blocks I had created were only visible for sites in the `00_Global` group.

I’m wondering if the reason you are getting unexpected behavior is because you’re not actually letting that “Resaving Matrix Blocks” job finish before you dive in to start editing your entries. If you don’t, then when you edit an entry, the Matrix blocks will still exist in all sites, and when you manually resave them in a particular site, that is the moment that the Propagation Method setting will kick in, removing the blocks from any sites that aren’t in the same group, or removing blocks from all other sites, depending on the setting value you choose in step 4.

To clarify, that setting is enforced whenever a Matrix block is saved. The Elements service’s `saveElement()` method will ensure that the block exists in all the sites it’s supposed to, and also make sure it is removed from any sites that it’s not supposed to be in. So when you change the Propagation Method setting value, the Matrix field will initiate a “Resaving Matrix Blocks” job, which loops through each Matrix block created for that field, and simply resaves it, and let the `saveElement()` method do its thing to apply the new Propagation Method selection.

So if you happen to manually save some Matrix blocks *before* the “Resaving Matrix Blocks” job gets to them, then you are the one who will be triggering the Propagation Method enforcement for those blocks. By the time the job gets to them, there won’t be anything left to do.

**Comment by elfacht** [comment]
@brandonkelly Thanks for your answer. I always let the task finish and sometimes, on localhost, the task failed so I ran it via `resave/matrix-blocks` – with the errors shown above. Most of the time the task runs as expected via the CP. As I said, the fields disappear before I save the entry. They're already gone when I open the edit page – after the resaving is done.

So my client has to stick with the fact, that there won't be any individual blocks in the future? This would be a major issue to run the website. Is there anything else I could try/do?

**Comment by brandonkelly** [comment]
Those `resave/matrix-blocks` errors are nothing to be worried about.

> So my client has to stick with the fact, that there won't be any individual blocks in the future? This would be a major issue to run the website. Is there anything else I could try/do?

I’m not sure what you mean by that?

**Comment by elfacht** [comment]
@brandonkelly What I meant: Is there anything I can do because the bug still occurs. Apperently it's an individual bug so better I send a support request via the CP, right?

**Comment by brandonkelly** [comment]
Sorry I had missed that you said that on localhost the “Resaving Matrix Blocks” task fails. So yeah sure, send a database backup and `composer.json` + `composer.lock` files over to support@craftcms.com and we can look into why it’s failing for you.

**Comment by elfacht** [comment]
@brandonkelly I finally could troubleshoot the bug, but can't say what's causing it. 

Steps to reproduce:

1. Create multiple sites with multiple languages. I did it like this: COM_DE, COM_EN, DE_DE, DE_EN
2. Put the sites in groups, just any groups, like `00_Global` and `Europe`.
3. Create a matrix field, set the propagation to **Save blocks to all sites the owner element is saved in** – a simple text block should do the trick.
4. Create an entry with the field and create a block. Save it to all sites.
5. Now: Rename the groups, move sites to other groups. I did it like this: Renamed `Europe` to `Germany`, moved DE_DE and DE_EN from `00_Global` to `Germany`.
6. Change the propagation method of the Matrix field to **Save blocks to other sites in the same site group**
7. Edit the entry, save it or not. Blocks on the sites that where moved to the renamed group are missing.

I've tested it with a clean installed Craft CMS. Once you re-organize your groups and move sites to other groups, the Matrix propagation will fail and will remove all blocks from the moved sites.

**Comment by jing-liu-anne** [comment]
@brandonkelly 

I have the same issue when using a draft after upgrading to Craft Pro 3.7.0.

**Steps to reproduce:**

1. Starts edits from Current and save as a draft.
2. Starts edits on the same field from Current and publishes them.
3. Switch to the draft version, the issue will appear.


https://user-images.githubusercontent.com/88373506/128153257-bd9aa471-0c9c-4805-8b29-ebb7467d7a06.mp4



I found it only happened on existing entry (I created them on Craft Pro 3.6.0.1 version or before). when I create a new entry now, the issue will not happen.

Thanks in advance!
